cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConfigurer.java

```diff
@@ -124,26 +133,139 @@ private WorkflowNode createNodeWithId(WorkflowNode node) {
+  private WorkflowNode createActionNodeWithId(WorkflowNode currentNode, WorkflowNode previousNode) {
+    WorkflowActionNode actionNode = (WorkflowActionNode) currentNode;
+    ScheduleProgramInfo program = actionNode.getProgram();
+    Set<String> parentNodeIds = new HashSet<>();
+    populateParentNodeIds(previousNode, parentNodeIds, currentNode);
+    if (program.getProgramType() == SchedulableProgramType.CUSTOM_ACTION) {
+      return new WorkflowActionNode(currentNode.getNodeId(), actionNode.getActionSpecification(), parentNodeIds);
+    }
+    return new WorkflowActionNode(currentNode.getNodeId(), program, parentNodeIds);
+  }
+
+  private WorkflowNode createForkNodeWithId(WorkflowNode currentNode, WorkflowNode previousNode) {
     String forkNodeId = Integer.toString(nodeIdentifier++);
     List<List<WorkflowNode>> branches = Lists.newArrayList();
-    WorkflowForkNode forkNode = (WorkflowForkNode) node;
+    WorkflowForkNode forkNode = (WorkflowForkNode) currentNode;
 
     for (List<WorkflowNode> branch : forkNode.getBranches()) {
-      branches.add(createNodesWithId(branch));
+      branches.add(createNodesWithId(branch, previousNode));
     }
     return new WorkflowForkNode(forkNodeId, branches);
   }
 
-  private WorkflowNode createConditionNodeWithId(WorkflowNode node) {
-    String conditionNodeId = node.getNodeId();
-    WorkflowConditionNode conditionNode = (WorkflowConditionNode) node;
+  private WorkflowNode createConditionNodeWithId(WorkflowNode currentNode, WorkflowNode previousNode) {
+    String conditionNodeId = currentNode.getNodeId();
+    WorkflowConditionNode conditionNode = (WorkflowConditionNode) currentNode;
     List<WorkflowNode> ifbranch = Lists.newArrayList();
     List<WorkflowNode> elsebranch = Lists.newArrayList();
-    ifbranch.addAll(createNodesWithId(conditionNode.getIfBranch()));
-    elsebranch.addAll(createNodesWithId(conditionNode.getElseBranch()));
+    ifbranch.addAll(createNodesWithId(conditionNode.getIfBranch(), conditionNode));
+    elsebranch.addAll(createNodesWithId(conditionNode.getElseBranch(), conditionNode));
+
+    Set<String> parentNodeIds = new HashSet<>();
+    populateParentNodeIds(previousNode, parentNodeIds, currentNode);
+    return new WorkflowConditionNode(conditionNodeId, conditionNode.getPredicateClassName(), ifbranch, elsebranch,
+                                     parentNodeIds);
+  }
+
+  // This method populates the set parentNodeIds with the parent node ids of the current node.
+  // If the previous node is of type ACTION, then that node is considered as the parent of the current node.
+  // If the previous node if of type FORK, then the last nodes on all the branches of FORK are considered as
+  // the parent of the current node.
+  // If the previous node is of type CONDITION and if the node is immediate child of the CONDITION, then the
+  // CONDITION node is considered to be parent of the current node, otherwise last nodes on the if branch and
```

Comment:
to be the parent of the current node; otherwise, the last nodes on the 'if' branch and
'else' branches are considered to be the parent of the current node.